File: Competitive Programming 3/Section 1.4/Palidromes/11309 - Counting Chaos/Python/main.py
def toInteger(h,m):
    return h*60 + m
def toSingleString(x):
    return ('0' if x<10 else '') + str(x)
def toString(h,m):
    return toSingleString(h)+toSingleString(m)
def toFString(h,m):
    return toSingleString(h)+':'+toSingleString(m)
def isPalindrome(h,m):
    s = toString(h,m)
    if s[0]=='0':
        s = s[next((i for i,val in enumerate(s)  if val!='0'),len(s)):]
    for i in range(len(s) >> 1):
        if s[i]!=s[len(s)-1-i]:
            return False
    return True
def main():
    nextPal = [''] * 1440
    nextP = '00:00'
    for h in range(23,-1,-1):
        for m in range(59,-1,-1):
            #print('what: ',h,m)
            nextPal[toInteger(h,m)]=nextP
            if isPalindrome(h,m):
                nextP=toFString(h,m)
    for _ in range(int(input())):
        s = input()
        h,m = [ int(s[i-1:i+1]) for i in range(1,len(s),3) ]
        print(nextPal[toInteger(h,m)])

main()

File: Competitive Programming 3/Section 1.4/Palidromes/11309 - Counting Chaos/Java/Main.java
import java.util.Scanner;

public class Main{
    static Scanner sc = new Scanner(System.in);
    static String[] nextPal = new String [1440];
    static int toInteger(int h,int m){
        return h*60 + m;
    }
    static String toSingleString(int x){
        return String.format("%02d",x);
    }
    static String toString(int h,int m){
        return toSingleString(h)+toSingleString(m);
    }
    static String toFString(int h,int m){
        return  toSingleString(h)+":"+ toSingleString(m);
    }
    static boolean isPalindrome(int h,int m){
        String s = toString(h,m);
        if(s.charAt(0)=='0')
            s = s.replaceFirst("0+","");
        for(int i=0; i < s.length();i++)
            if(s.charAt(i)!=s.charAt(s.length()-1-i))
                return false;
        return true;
    }
    public static void main(String[]args){
        String nextP = "00:00";
        for(int h = 23 ; h > -1 ; h --){
            for (int m = 59 ; m > -1 ; m --){
                nextPal[toInteger(h,m)]=nextP;
                if(isPalindrome(h,m))
                    nextP=toFString(h,m);
            }
        }
        for(int t = sc.nextInt(); t > 0 ; t --){
            String time = sc.next();
            int h=Integer.parseInt(time.substring(0,2)),m=Integer.parseInt(time.substring(3,5));
            System.out.printf("%s\n",nextPal[toInteger(h,m)]);
        }
    }
}